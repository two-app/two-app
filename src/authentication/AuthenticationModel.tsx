export type Tokens = {
  accessToken: string;
  refreshToken: string;
};

export type UserRegistration = {
  uid: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  dob: string;
  acceptedTerms: boolean;
};
