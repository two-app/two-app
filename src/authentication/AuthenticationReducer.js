// @flow

import {createAction, createReducer} from "@reduxjs/toolkit";

const setTokens = createAction("SET_TOKENS");

const authenticationReducer = createReducer(null, {
    [setTokens.type]: (state, action) => ({
        ...state,
        accessToken: action.payload['access_token'],
        refreshToken: action.payload['refresh_token']
    })
});

export default authenticationReducer;
export {setTokens};