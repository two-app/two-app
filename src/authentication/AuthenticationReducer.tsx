import {createAction, createReducer} from '@reduxjs/toolkit';

const setTokens = createAction('SET_TOKENS');

const authenticationReducer = createReducer(null, {
    [setTokens.type]: (state: any, action) => ({
        ...state,
        accessToken: action.payload['accessToken'],
        refreshToken: action.payload['refreshToken']
    })
});

export default authenticationReducer;
export {setTokens};