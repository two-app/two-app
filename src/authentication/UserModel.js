// @flow

import jwt_decode from "jwt-decode";

const UnconnectedUser: {
    uid: number,
    connectCode: string
} = {
    uid: null,
    connectCode: null
};

const createUnconnectedUser = (uid: number, connectCode: string): UnconnectedUser => ({uid, connectCode});

const parseUnconnectedUserFromToken = (accessToken: string): (UnconnectedUser) => {
    const decoded = jwt_decode(accessToken);
    const uid = decoded['uid'];
    const connectCode = decoded['connectCode'];

    return createUnconnectedUser(uid, connectCode);
};

export {UnconnectedUser, createUnconnectedUser, parseUnconnectedUserFromToken};