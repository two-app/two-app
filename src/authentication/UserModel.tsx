import jwt_decode from 'jwt-decode';

// @ts-ignore
const decode = (token: string): any => jwt_decode(token);

class UnconnectedUser {
    uid: number;
    connectCode: string;

    constructor(uid: number, connectCode: string) {
        this.uid = uid;
        this.connectCode = connectCode;
    }
    // @ts-ignore
    static fromAccessToken = (accessToken: string) => new UnconnectedUser(
        decode(accessToken)['uid'],
        decode(accessToken)['connectCode']
    );
}

class User {
    uid: number;
    pid: number;
    cid: number;

    constructor(uid: number, pid: number, cid: number) {
        this.uid = uid;
        this.pid = pid;
        this.cid = cid;
    }

    static fromAccessToken = (accessToken: string) => new User(
        decode(accessToken)['uid'],
        decode(accessToken)['pid'],
        decode(accessToken)['cid']
    );
}

export {UnconnectedUser, User};