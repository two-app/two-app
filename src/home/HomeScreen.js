import React from "react";
import {connect} from "react-redux";
import {WrapperContainer} from "../views/View";
import {Text, TouchableOpacity} from "react-native";
import {NavigationActions, StackActions} from "react-navigation";

/**
 * @param user {User}
 */
const HomeScreen = ({navigation, user}) => {
    const logout = () => navigation.dispatch(
        StackActions.reset({
            index: 0,
            actions: [NavigationActions.navigate({routeName: 'LogoutScreen'})]
        })
    );

    return <WrapperContainer>
        <Text>You're logged in.</Text>
        <Text>Your UID: {user.uid}</Text>
        <Text>Your PID: {user.pid}</Text>
        <Text>Your CID: {user.cid}</Text>
        <TouchableOpacity onPress={logout}><Text>Logout</Text></TouchableOpacity>
    </WrapperContainer>;
};

HomeScreen.navigationOptions = {
    title: 'Home',
    header: null
};

const mapStateToProps = state => ({user: state['user']});

export default connect(mapStateToProps)(HomeScreen);
export {HomeScreen};