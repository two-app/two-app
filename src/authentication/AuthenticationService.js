// @flow
import Gateway from "../http/Gateway";
import {AxiosError, AxiosResponse} from "axios";
import {UserRegistration} from "./register_workflow/UserRegistrationModel";
import {UnconnectedUser} from "./UserModel";
import {Tokens} from "./AuthenticationModel";

class RegisterUserResponse {
    user: UnconnectedUser;
    tokens: Tokens;

    constructor(user: UnconnectedUser, tokens: Tokens) {
        this.user = user;
        this.tokens = tokens;
    }
}


const registerUser = (userRegistration: UserRegistration): Promise<RegisterUserResponse> =>
    Gateway.post("/self", userRegistration)
        .then((r: AxiosResponse) => {
            const accessToken = r.data['accessToken'];
            const refreshToken = r.data['refreshToken'];
            return new RegisterUserResponse({
                user: UnconnectedUser.fromAccessToken(accessToken),
                tokens: new Tokens(accessToken, refreshToken)
            });
        }).catch((e: AxiosError) => {
            throw new Error(e.response.data['message'].toString());
        }
    );

export default {registerUser};
export {RegisterUserResponse};