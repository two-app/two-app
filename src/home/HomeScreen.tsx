import {StackNavigationProp} from '@react-navigation/stack';
import React from 'react';
import {connect} from 'react-redux';
import {RootStackParamList} from '../../Router';
import {User} from '../authentication/UserModel';
import {Memories} from '../memories/Memories';
import {NoWrapContainer, Wrapper} from '../views/View';
import {Footer} from './Footer';

type HomeScreenProps = {
  navigation: StackNavigationProp<RootStackParamList, 'HomeScreen'>;
  user: User;
};

const HomeScreen = ({navigation, user}: HomeScreenProps) => (
  <Wrapper>
    <NoWrapContainer>
      <Memories />
    </NoWrapContainer>
    <Footer active="HomeScreen" />
  </Wrapper>
);

const mapStateToProps = (state: any) => ({user: state['user']});

export default connect(mapStateToProps)(HomeScreen);
export {HomeScreen};
