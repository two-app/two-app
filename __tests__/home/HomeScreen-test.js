// @flow

import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {HomeScreen} from "../../src/home/HomeScreen";
import {User} from "../../src/authentication/UserModel";

describe('HomeScreen', () => {
    const user = new User(1, 2, 3);

    test('should maintain snapshot', () => expect(renderer.create(<HomeScreen user={user}/>)).toMatchSnapshot());
});
