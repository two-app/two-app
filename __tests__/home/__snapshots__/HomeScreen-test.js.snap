// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should maintain snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": "10%",
            "marginRight": "10%",
          }
        }
      >
        <Text>
          You're logged in. Home screen, booyah!
        </Text>
        <Text>
          Your UID: 
          1
        </Text>
        <Text>
          Your PID: 
          2
        </Text>
        <Text>
          Your CID: 
          3
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
